In [23]:
import numpy as np
import pandas as pd
from pandas_profiling import ProfileReport
import matplotlib.pyplot as plt

## Exploratory data analysis

The inspection of data is an essential step in the data analysis process, as it enables a comprehensive understanding of the characteristics of the dataset, identification of any inconsistencies or issues.

There are a variety of methods that can be employed to inspect data, such as:

* Examining the data types: It is essential to ensure that all variables are represented in the appropriate data type. This includes ensuring numerical variables are stored as integers or floats, and categorical variables are stored as strings or factors.

* Examination of missing values: A thorough examination for missing or null values in the dataset is necessary. Missing values can pose challenges when building machine learning models, thus it is crucial to identify and handle them appropriately.

* Analysis of data distribution: Visualization of the distribution of variables through histograms or box plots is a valuable method to identify outliers or skewed distributions that may require special handling or transformation.

* Examining the relationships between variables: Scatterplots or correlation analysis can be used to identify relationships between variables. This can assist in identifying potential multicollinearity/redundancy problems with variables.

This notebooks is dedicated for data exploration where i will create data profiling reports for train and test datasets and comparison between them.

In [12]:
train = pd.read_csv('../data/train.csv')
test = pd.read_csv('../data/test.csv')

In [13]:
numerical_data = train.select_dtypes("number")
numerical_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 38 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   LotFrontage    1201 non-null   float64
 3   LotArea        1460 non-null   int64  
 4   OverallQual    1460 non-null   int64  
 5   OverallCond    1460 non-null   int64  
 6   YearBuilt      1460 non-null   int64  
 7   YearRemodAdd   1460 non-null   int64  
 8   MasVnrArea     1452 non-null   float64
 9   BsmtFinSF1     1460 non-null   int64  
 10  BsmtFinSF2     1460 non-null   int64  
 11  BsmtUnfSF      1460 non-null   int64  
 12  TotalBsmtSF    1460 non-null   int64  
 13  1stFlrSF       1460 non-null   int64  
 14  2ndFlrSF       1460 non-null   int64  
 15  LowQualFinSF   1460 non-null   int64  
 16  GrLivArea      1460 non-null   int64  
 17  BsmtFullBath   1460 non-null   int64  
 18  BsmtHalf

In [14]:
categorical = train.select_dtypes(object)
categorical.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 43 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   MSZoning       1460 non-null   object
 1   Street         1460 non-null   object
 2   Alley          91 non-null     object
 3   LotShape       1460 non-null   object
 4   LandContour    1460 non-null   object
 5   Utilities      1460 non-null   object
 6   LotConfig      1460 non-null   object
 7   LandSlope      1460 non-null   object
 8   Neighborhood   1460 non-null   object
 9   Condition1     1460 non-null   object
 10  Condition2     1460 non-null   object
 11  BldgType       1460 non-null   object
 12  HouseStyle     1460 non-null   object
 13  RoofStyle      1460 non-null   object
 14  RoofMatl       1460 non-null   object
 15  Exterior1st    1460 non-null   object
 16  Exterior2nd    1460 non-null   object
 17  MasVnrType     1452 non-null   object
 18  ExterQual      1460 non-null

In [16]:
#Feature separation:
discrete = ['YearBuilt', 'YearRemodAdd','BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath',
            'BedroomAbvGr', 'KitchenAbvGr','TotRmsAbvGrd','Fireplaces', 'GarageYrBlt','GarageCars', 
            'MoSold', 'YrSold', 'OverallQual', 'OverallCond']

continuous = ['LotFrontage', 'LotArea','MasVnrArea','BsmtFinSF1',  'BsmtFinSF2', 'BsmtUnfSF', 
              'TotalBsmtSF','1stFlrSF', '2ndFlrSF', 'LowQualFinSF','GrLivArea', 'GarageArea',  
              'WoodDeckSF', 'OpenPorchSF', 'EnclosedPorch', '3SsnPorch','ScreenPorch', 'PoolArea', 
              'MiscVal']

nominal = ['MSSubClass','MSZoning', 'Alley',  'LandContour','LotConfig',   'Neighborhood', 
           'Condition1', 'Condition2', 'BldgType', 'HouseStyle','RoofStyle','RoofMatl', 'Exterior1st',
           'Exterior2nd', 'MasVnrType',  'Foundation','Heating',  'CentralAir',  'GarageType','MiscFeature',
           'SaleType', 'SaleCondition']

ordinal = ['LotShape', 'LandSlope',  'ExterQual', 'ExterCond', 'BsmtQual', 'BsmtCond',
           'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2','HeatingQC', 'Electrical','KitchenQual', 
           'Functional','FireplaceQu', 'GarageFinish', 'GarageQual','GarageCond','PavedDrive', 
           'PoolQC', 'Fence']

In [22]:
# Train report
train_profile = ProfileReport(
    train, 
    title="EDA report for training dataset",
    correlations={
        "phi_k": {"calculate": True},
    },
    )
train_profile.to_notebook_iframe()
train_profile.to_file("../reports/train_dataframe_eda.html")

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

KeyboardInterrupt: 

In [ ]:
# Test report

test_profile = ProfileReport(
    test, 
    title="EDA report for test dataset",
    correlations={
        "phi_k": {"calculate": True},
        },
    )
test_profile.to_notebook_iframe()
#train_profile.to_widgets()
test_profile.to_file("../reports/test_dataframe_eda.html")

In [ ]:
# Comparison of test and train reports
comparison_report = train_profile.compare(test_profile)
comparison_report.to_file("comparison.html")

### Key concludions

The target variable, sale price, reveals a right-skewed distribution. This indicates that a small proportion of data points exhibit a much higher value compared to the majority of the observations. This can be visualized in the histogram, where the peak is inclined towards the left, and there is a long tail towards the right. The right-skewness implies that the house prices are not normally distributed, as it deviates from the bell-shaped curve. Such deviation can have an impact on the results of certain statistical methods.

It is acknowledged that skewness in real-world data is not an uncommon phenomenon. To mitigate the effect of skewness, various mathematical techniques, such as power transforms, including logarithmic, square root, and reciprocal transformations, can be applied to the data to make it conform to a normal distribution.

The dataset is heterogeneous, comprising both numeric and categorical data types. In total, the dataset comprises 79 features and 1460 samples. However, it should be noted that certain features within the dataset have missing values.

A key observation is that some features appears to exhibit high levels of correlation, which raises concerns regarding multicollinearity. Multicollinearity refers to the scenario where two or more predictor variables are highly correlated, and as a result decrease models stability. This will be examined in further sections of this notebook.

Furthermore, some features may have low variance, indicating that the values for these features do not vary significantly across the dataset. These features may not contribute new information and their presence will not improve predictive power. 

Many of the numerical features are represented with integer values, which means that these features can be divided into continuous and discrete categories. It is important to note that some features, such as OverallQl, have been pre-encoded as ordinal object features.

Similar conclusions can be drawn when examining the categorical variables in the dataset; some exhibit skewness, while others have low variance.

#### Detailed look at missing values

In [ ]:
missing = train.isnull().sum()
missing = missing[missing > 0]
missing.plot.bar()
plt.xlabel("Features")
plt.ylabel("Number of missing values")
plt.title("Missing values")
plt.show()

In [ ]:
numerical_data.loc[:,numerical_data.isnull().sum()>0].head(5)

In [ ]:
train[nominal].loc[:,train[nominal].isnull().sum()>0].head(5)

In [ ]:
train[ordinal].loc[:,train[ordinal].isnull().sum()>0].head(5)

It can be inferred that the majority of missing values observed in the dataset are a result of the absence of the respective feature in the given sample. This is evident in the case of numeric and nominal variables such as 'LotFrontage', 'MasVnrArea', 'GarageYrBlt' 'Alley', 'MasVnrType','GarageType', 'MiscFeature, as well as for ordinal variables such as 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2'. However, the missing value for one row in the variable 'Electrical' is peculiar and would require further examination or exclusion of this sample.

In [ ]:
discrete = ['YearBuilt', 'YearRemodAdd','BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath',
            'BedroomAbvGr', 'KitchenAbvGr','TotRmsAbvGrd','Fireplaces', 'GarageYrBlt','GarageCars', 
             'OverallQual', 'OverallCond']

continuous = ['LotFrontage', 'LotArea','MasVnrArea','BsmtFinSF1',  'BsmtFinSF2', 'BsmtUnfSF', 
              'TotalBsmtSF','1stFlrSF', '2ndFlrSF', 'LowQualFinSF','GrLivArea', 'GarageArea',  
              'WoodDeckSF', 'OpenPorchSF', 'EnclosedPorch','ScreenPorch', 'PoolArea' 
              ]

nominal = ['MSSubClass','MSZoning', 'Street', 'Alley',  'LandContour','LotConfig',   'Neighborhood', 
           'Condition1', 'Condition2', 'BldgType', 'HouseStyle','RoofStyle','RoofMatl', 'Exterior1st',
           'Exterior2nd', 'MasVnrType',  'Foundation','Heating',  'CentralAir',  'GarageType',
           'SaleType', 'SaleCondition']

ordinal = ['LotShape', 'LandSlope', 'Utilities',   'ExterQual', 'ExterCond', 'BsmtQual', 'BsmtCond',
           'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2','HeatingQC', 'Electrical','KitchenQual', 
           'Functional','FireplaceQu', 'GarageFinish', 'GarageQual','GarageCond','PavedDrive', 
           'PoolQC', 'Fence']

#potentially

drop = ['Id', 'Street', 'Utilities',"Condition2", '3SsnPorch','MiscFeature','MiscVal', 'MoSold', 'YrSold'], 

pca = ['LandContour', 'LandSlope', "Neighborhood"]

pca = ['RoofStyle','RoofMatl']

pca = ['BsmtQual','BsmtCond','BsmtExposure','BsmtFinType1','BsmtUnfSF']

pca = ['Condition1', 'Condition2',]

impute_most_frequent = ['Electrical']


ply = ['BsmtFinSF1', 'TotalBsmtSF', '1stFlrSF', 'GrLivArea', 'GarageArea']